### PR TITLE
33 refactor code to use routers

### LINKED ISSUE #33: Refactor code to use routers
Sounds like a plan! Just trying to get this POC completed so i can begin to work on the MVP.

_Originally posted by @Tanner-Ray-Martin in https://github.com/Tanner-Ray-Martin/Dynamic-DB/pull/32#discussion_r1714058704_

### PR DESCRIPTION
**Pull request by Tanner-Ray-Martin**
split main api up into routes
fixed validation bug
updated route path structure
added tags route and models

### CODE REVIEW

File: POC/api/main.py [hunk 1]
```diff
@@ -1,655 +1,32 @@
...
-                            loading=[c.Text(text="Submitting")],
-                            model=DbInfoForm,
-                        ),
-                    ]
-                ),
-                c.Markdown(text="---"),
-            ]
-        )
-    ]
+from fastapi import APIRouter, FastAPI
```
Comment (Tanner-Ray-Martin) [inline POC/api/main.py:1]: import the api router and new routes

File: POC/api/main.py [hunk 2]
```diff
@@ -1,655 +1,32 @@
...
 
+from POC.api.routes.forms import field_forms, db_forms, tag_forms
+from POC.api.routes.backend import field_apis, db_apis
+from POC.api.routes import base
 
-@app.post(
-    "/api/forms/update/database/{database_id}",
-    response_model=FastUI,
-    response_model_exclude_none=True,
+app = FastAPI(title="Dynamic-DB")
```
Comment (Tanner-Ray-Martin) [inline POC/api/main.py:7]: included the backend "/api/" and fastui "/forms" to the api router

File: POC/api/main.py [hunk 3]
```diff
@@ -1,655 +1,32 @@
...
-            submit_url=f"/api/forms/create/databaseField/{database_id}",
-        ),
-    ]
-
-
-@app.post(
-    "/api/forms/create/databaseField/{database_id}",
-    response_model=FastUI,
-    response_model_exclude_none=True,
+api_router.include_router(
```
Comment (Tanner-Ray-Martin) [inline POC/api/main.py:23]: included the newly created tags route to router

File: POC/api/main.py [hunk 4]
```diff
@@ -1,655 +1,32 @@
...
-    """Simple HTML page which serves the React app, comes last as it matches all paths."""
-    return HTMLResponse(prebuilt_html(title="FastUI Demo"))
-
 
-if __name__ == "__main__":
-    import uvicorn
+api_router.include_router(base.router, tags=["base"], include_in_schema=False)
 
-    uvicorn.run(app, reload=True)
+app.include_router(api_router)
```
Comment (Tanner-Ray-Martin) [inline POC/api/main.py:32]: added the router to the FastApi app

File: POC/api/routes/backend/db_apis.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+from typing import Sequence
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/backend/db_apis.py:1]: routes for the api portion of the database creation

File: POC/api/routes/backend/field_apis.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+from typing import Sequence
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/backend/field_apis.py:1]: routes for the api portion of the field creation

File: POC/api/routes/base.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+from fastui.events import GoToEvent
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/base.py:1]: fastui and basic routes like redirects and setting the fastui base route along with the api value for the backend portion of the fastui handler

File: POC/api/routes/base.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    return RedirectResponse("/openapi.json", status_code=302)
+
+
+@router.get("/", response_class=RedirectResponse)
+async def home_redirect() -> RedirectResponse:
+    return RedirectResponse("/welcome", status_code=302)
+
+
+@router.get("/{path:path}")
+async def html_landing() -> HTMLResponse:
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/base.py:85]: set the api root url to "/forms" so anything with /api is used for the actual api and not the fastui api.
"/databases/read/1" FastUI HTMLResponse
"/forms/databases/read/1" FastUI JSONResponse
"/api/databases/read/1" regular api JSONResponse

File: POC/api/routes/forms/db_forms.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
+from typing import Annotated
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/db_forms.py:1]: routes for db fastui responses

File: POC/api/routes/forms/field_forms.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
+from typing import Annotated
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/field_forms.py:1]: routes for field fastui responses

File: POC/api/routes/forms/tag_forms.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations as _annotations
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/tag_forms.py:1]: routes for tags fastui responses

File: POC/db/models/stock_models/db_models.py [hunk 12]
```diff
@@ -2,18 +2,17 @@
 from sqlmodel import SQLModel, Field, create_engine
 
 from datetime import datetime as dt, timedelta, date
-from typing import Literal, Optional, Any
+from typing import Literal, Optional
 from pydantic import (
     BaseModel,
-    # field_validator,
```
Comment (Tanner-Ray-Martin) [inline POC/db/models/stock_models/db_models.py:8]: cleaned up imports

File: POC/db/models/stock_models/db_models.py [hunk 13]
```diff
@@ -23,50 +22,47 @@
...
+        description="Write a detailed description of this database and its purpose.",
     )
 
 
-class DbInfo(DbInfoForm):  # type: ignore
+# Model Definition That performs Validation
+class DbInfo(DbInfoForm):
     id: Optional[int] = Field(default=None, primary_key=True)
-    created_at: dt = Field(default=dt.now())
+    created_at: dt = Field(default_factory=dt.now)
```
Comment (Tanner-Ray-Martin) [inline POC/db/models/stock_models/db_models.py:56]: not sure why but default factory for all dt fields works better

File: POC/db/models/stock_models/db_models.py [hunk 14]
```diff
@@ -23,50 +22,47 @@
...
-    @classmethod
-    def remove_ms_from_dt(cls, data: Any) -> Any:
-        if isinstance(data, dict):
-            for field_name, field_value in data.items():
-                if isinstance(field_value, dt):
-                    data.update({field_name: field_value.replace(microsecond=0)})
-            else:
-                assert isinstance(data, dict)
-        return data
+    @field_serializer("created_at", "updated_at", when_used="always")
```
Comment (Tanner-Ray-Martin) [inline POC/db/models/stock_models/db_models.py:60]: changed dt validation to occur when model dump is called

File: POC/db/models/stock_models/db_models.py [hunk 15]
```diff
@@ -106,15 +102,43 @@ class FieldInfoForm(SQLModel):
 class FieldInfo(FieldInfoForm):
     id: int = Field(default=None, primary_key=True)
     db_id: int = Field(default=None)
-    created_at: dt = Field(default=dt.now())
-    updated_at: dt = Field(default=dt.now())
+    created_at: dt = Field(default_factory=dt.now)
```
Comment (Tanner-Ray-Martin) [inline POC/db/models/stock_models/db_models.py:105]: added dt field field_serialization to occur when calling model dump

File: POC/db/models/stock_models/db_models.py [hunk 16]
```diff
@@ -106,15 +102,43 @@ class FieldInfoForm(SQLModel):
...
+class TagsInfo(CreateTagsForm, table=True):  # type: ignore
+    id: Optional[int] = Field(default=None, primary_key=True)
+
+
+class SelectTagForm(SQLModel):
+    tag_name: list[str] = PydanticField(
+        json_schema_extra={"search_url": "/forms/tags/search"}
+    )
+
+    @field_validator("tag_name", mode="before")
```
Comment (Tanner-Ray-Martin) [inline POC/db/models/stock_models/db_models.py:134]: this is a form for selection on or more tags.
validates that no matter what the users response is the selected tags or tag is a list

File: POC/tests/test_api.py [hunk 17]
```diff
@@ -43,21 +43,21 @@ def db_field_form() -> FieldInfoForm:
 
 @pytest.fixture
 def frontend_url() -> str:
-    return "/api/forms/"
```
Comment (Tanner-Ray-Martin) [inline POC/tests/test_api.py:46]: updated all routes in tests to the new structure

File: pyproject.toml [hunk 18]
```diff
@@ -34,6 +34,7 @@ mypy = ">=1.11.1,<1.12"
 test = "pytest"
 pre_commit = "pre-commit run --all-files"
 mypy_poc = "mypy POC --check-untyped-defs --ignore-missing-imports"
+poc = "uvicorn POC.api.main:app --reload"
```
Comment (Tanner-Ray-Martin) [inline pyproject.toml:37]: added poc as an option in test environment with --reload

File: POC/api/routes/forms/tag_forms.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    db_engine,
+    SelectTagForm,
+    CreateTagsForm,
+)
+
+router = APIRouter()
+
+
+@router.get("/search", response_model=SelectSearchResponse)
+async def search_view() -> SelectSearchResponse:
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/tag_forms.py:23]: the tag selection option are pulled from a database so the user can select them

File: POC/api/routes/forms/tag_forms.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        tags: list[SelectOption] = [
+            SelectOption(value=tag.tag_name, label=tag.tag_name) for tag in data
+        ]
+        # options = [{"label": k, "options": v} for k, v in tags.items()]
+
+    return SelectSearchResponse(options=tags)
+
+
+@router.get("/select", response_model=FastUI, response_model_exclude_none=True)
+def select_tags():
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/tag_forms.py:38]: the form where the user can select multiple tags

File: POC/api/routes/forms/tag_forms.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+                    display_mode="page",
+                    submit_url="/forms/tags/select",
+                ),
+            ]
+        )
+    ]
+
+
+@router.post("/select", response_model=FastUI, response_model_exclude_none=True)
+async def submit_selected_tags(
```
Comment (Tanner-Ray-Martin) [inline POC/api/routes/forms/tag_forms.py:58]: test functionality that the tag selection works.
displays a table of the selected tags

**Review by Tanner-Ray-Martin — COMMENTED** [review]
@Zaubeerer 
Can you check this out?

**Comment by Zaubeerer** [comment]
@Tanner-Ray-Martin, I will have a look at this later today :)

**Review by Zaubeerer — APPROVED** [review]
Sorry for the delay, looks good to me :)

Great how routers help to make the code more readable, right?

**Pull request merged**